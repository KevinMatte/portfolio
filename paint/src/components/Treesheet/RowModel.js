import {getValueByPath} from "../../general/Utils";
import {cloneObject} from "../../redux/PathUtils";

export default class RowModel {

    constructor(props) {
        this.props = props;
        this.sheetsByName = {};
        this.sheetNames = [];
        this.rows = [];
        this.numColumns = 0;
        this.columns = [];

        this.createRows(this.props.dataTree);
        this.updateSpreadsheetOpenRows();
    }

    createRow(obj, path) {
        let {types} = this.props;

        let typeName = obj['type'];
        if (!typeName || !types.hasOwnProperty(typeName))
            typeName = obj['table'];
        let type = types[typeName];
        if (!type)
            return;

        let sheetName = `${typeName}_${path.length}`;

        let values = type.columns.map(column => {
            return getValueByPath(obj, column.path);
        });
        this.numColumns = Math.max(this.numColumns, type.columns.length);
        if (!this.sheetsByName.hasOwnProperty(sheetName)) {
            this.sheetsByName[sheetName] = {
                typeName,
                sheetName,
                path,
                type,
            };
            this.sheetNames.push(sheetName);
        }
        let row = {
            typeName,
            sheetName,
            values,
            path,
        };
        this.rows.push(row);

        if (type.hasOwnProperty('fields')) {
            type.fields.every((key) => {
                this.createRow(obj[key], [...path, key]);
                row.isOpen = true;
                return true;
            })
        }

        if (type.hasOwnProperty('arrays')) {
            type.arrays.every((key) => {
                this.createRows(obj[key], [...path, key]);
                row.isOpen = true;
                return true;
            })
        }
    }

    createRows(aList, path = []) {
        aList.every((obj, iObj) => {
            this.createRow(obj, [...path, iObj]);
            return obj;
        });
    }

    duplicateRow(iFirstRow) {
        let iRow = iFirstRow;
        let iSplicePoint = iFirstRow;
        let newRows = [];
        let indent = this.rows[iRow].path.length;
        while (iRow < this.rows.length && (iRow === iFirstRow || this.rows[iRow].path.length > indent)) {
            let newRow = cloneObject(this.rows[iRow]);
            newRow.path[indent - 1]++;
            newRows.push(newRow);
            iRow++;
        }
        if (newRows.length > 0) {
            iSplicePoint = iRow;
            while (iRow < this.rows.length && this.rows[iRow].path.length >= indent) {
                this.rows[iRow].path[indent - 1]++;
                iRow++;
            }
            this.rows.splice(iSplicePoint, 0, ...newRows);
        }
        this.updateSpreadsheetOpenRows();
        return iSplicePoint;
    }

    deleteRow(iFirstRow) {
        let iRow = iFirstRow;
        let iSplicePoint = iFirstRow;
        let deleteCount = 0;
        let indent = this.rows[iRow].path.length;
        while (iRow < this.rows.length && (iRow === iFirstRow || this.rows[iRow].path.length > indent)) {
            iRow++;
            deleteCount++;
        }
        while (iRow < this.rows.length && this.rows[iRow].path.length >= indent) {
            this.rows[iRow].path[indent - 1]--;
            iRow++;
        }
        this.rows.splice(iSplicePoint, deleteCount);

        this.updateSpreadsheetOpenRows();
        return iSplicePoint;
    }

    updateSpreadsheetOpenRows() {
        let openIndent = -1;
        this.openRows = this.rows.filter((row) => {
            let indent = row.path.length;
            if (openIndent !== -1 && indent > openIndent)
                return false;
            if (indent === openIndent)
                openIndent = -1;
            if (!row.isOpen)
                openIndent = indent;
            return true;
        });
    }
}

