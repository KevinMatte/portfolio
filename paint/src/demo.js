/* Copyright (C) 2019 Kevin Matte - All Rights Reserved */

import store from "./redux";
import TreeState from "./redux/treeState";

export let types = {
    drawing: {
        columns: [
            {label: 'Name', width: '150px', path: 'name'},
            {label: 'Type', width: '150px', path: 'type'},
        ],
        arrays: ['graphs'],
        fields: [],
    },
    graph: {
        columns: [
            {label: 'Name', width: '160px', path: 'name'},
            {label: 'Type', width: '160px', path: 'type'},
        ],
        arrays: ['points'],
        fields: [],
    },
    vector3: {
        columns: [
            {label: 'Name', width: '150px', path: 'name'},
            {label: 'Type', width: '150px', path: 'type'},
            {label: 'X', width: '150px', path: ['axis', 0]},
            {label: 'Y', width: '150px', path: ['axis', 1]},
            {label: 'Z', width: '150px', path: ['axis', 2]},
        ],
        array: [],
        fields: [],
    },
};
export let dataTree = [
    {
        name: 'First drawing',
        type: 'drawing',
        table: 'drawing',
        drawingId: 1,
        graphs: [
            {
                name: 'line 1',
                type: 'line',
                table: 'graph',
                points: [
                    {
                        type: 'point',
                        name: 'start',
                        table: 'vector3',
                        axis: [1, 2, 3],
                    },
                    {
                        type: 'point',
                        name: 'end',
                        table: 'vector3',
                        axis: [3, 2, 1],
                    },

                ]
            },
            {
                name: 'cube 1',
                type: 'cube',
                table: 'graph',
                points: [
                    {
                        type: 'point',
                        name: 'vertex',
                        table: 'vector3',
                        axis: [11, 21, 31],
                    },
                    {
                        type: 'vector',
                        name: 'edge',
                        table: 'vector3',
                        axis: [0, 0, 5],
                    },

                ]
            },
            {
                name: 'box 1',
                type: 'box',
                table: 'graph',
                points: [
                    {
                        table: 'vector3',
                        type: 'point',
                        name: 'vertex',
                        axis: [-11, 21, 31],
                    },
                    {
                        table: 'vector3',
                        type: 'vector',
                        name: 'x',
                        axis: [0, 0, 5],
                    },
                    {
                        table: 'vector3',
                        type: 'vector',
                        name: 'y',
                        axis: [0, 5, 0],
                    },
                    {
                        table: 'vector3',
                        type: 'vector',
                        name: 'z',
                        axis: [5, 0, 0],
                    },

                ]
            },
        ],
    },
];

export function setupDemo() {
    store.dispatch(TreeState.setTypes(types));
    store.dispatch(TreeState.setDrawings(dataTree));
}
