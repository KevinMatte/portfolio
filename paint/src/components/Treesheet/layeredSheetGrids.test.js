// Cell.react.test.js
import React from 'react';
import {render, fireEvent, cleanup, waitForElement} from 'react-testing-library';
import 'jest-dom/extend-expect'
import {LayeredSheetGrids, Controller} from './layeredSheetGrids';
import RowModel from "./rowModel";
import TreeModel from "../../redux/treeModel";
import {setupDemo, types, dataTree} from "../../demo";

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup);

// noinspection ES6ModulesDependencies
test('LayeredSheetGrids: Basic', () => {
    let tree, value;

    let name = "treesheet";
    let rowHeight = 50;
    let indentPixels = 30;
    let gridSpacingWidth = 5;

    let rowModel = new RowModel({name, types, dataTree});
    let commonProps = {name, rowHeight, gridSpacingWidth, rowModel};
    let colHeaderProps = {...commonProps, indentPixels};

    const {container, getByTestId, asFragment} = render(
        <LayeredSheetGrids {...colHeaderProps}/>
    );

    tree = asFragment();
    expect(tree).toMatchSnapshot();

});

