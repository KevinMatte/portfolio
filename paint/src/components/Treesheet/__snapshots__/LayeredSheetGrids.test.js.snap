// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayeredSheetGrids: Basic 1`] = `
<DocumentFragment>
  <div
    class="max_size overflowAuto"
  >
    <div
      class="SpreadsheetOverlay"
    >
      <div
        class="Spreadsheet max_size"
        id="drawing_1"
      >
        <div
          class="middleText"
        >
          <button
            data-testid="0"
          >
            -
          </button>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0:0"
          >
            First drawing
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0:1"
          >
            drawing
          </span>
        </div>
      </div>
      <div
        class="Spreadsheet max_size"
        id="graph_3"
      >
        <div
          class="middleText"
        >
          <button
            data-testid="0,graphs,0"
          >
            -
          </button>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0:0"
          >
            line 1
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0:1"
          >
            line
          </span>
        </div>
        <div
          class="middleText"
        >
          <button
            data-testid="0,graphs,1"
          >
            -
          </button>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1:0"
          >
            cube 1
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1:1"
          >
            cube
          </span>
        </div>
        <div
          class="middleText"
        >
          <button
            data-testid="0,graphs,2"
          >
            -
          </button>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2:0"
          >
            box 1
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2:1"
          >
            box
          </span>
        </div>
      </div>
      <div
        class="Spreadsheet max_size"
        id="vector3_5"
      >
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,0:0"
          >
            start
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,0:1"
          >
            point
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,0:2"
          >
            1
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,0:3"
          >
            2
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,0:4"
          >
            3
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,1:0"
          >
            end
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,1:1"
          >
            point
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,1:2"
          >
            3
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,1:3"
          >
            2
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,0,points,1:4"
          >
            1
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,0:0"
          >
            vertex
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,0:1"
          >
            point
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,0:2"
          >
            11
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,0:3"
          >
            21
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,0:4"
          >
            31
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,1:0"
          >
            edge
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,1:1"
          >
            vector
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,1:2"
          >
            0
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,1:3"
          >
            0
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,1,points,1:4"
          >
            5
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,0:0"
          >
            vertex
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,0:1"
          >
            point
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,0:2"
          >
            -11
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,0:3"
          >
            21
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,0:4"
          >
            31
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,1:0"
          >
            x
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,1:1"
          >
            vector
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,1:2"
          >
            0
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,1:3"
          >
            0
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,1:4"
          >
            5
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,2:0"
          >
            y
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,2:1"
          >
            vector
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,2:2"
          >
            0
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,2:3"
          >
            5
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,2:4"
          >
            0
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,3:0"
          >
            z
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,3:1"
          >
            vector
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,3:2"
          >
            5
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,3:3"
          >
            0
          </span>
        </div>
        <div
          class="ValueCell Cell middleText"
        >
          <span
            data-testid="0,graphs,2,points,3:4"
          >
            0
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
